### DISCUSSION TITLE
RfC - discuss possible use cases of LLM like chatGPT in the CMS context

### DISCUSSION THREAD

**Discussion by macolo** [django-cms/django-cms#7976] [General]
## Possible use cases

Examples below.

### Content generation
- LLM to create a page based on available plugins and content directions.
- LLM to populate abstract field for blog posts

### SEO
- LLM to propose SEO page titles and descriptions optimizations based on existing page content

**Comment by macolo** [comment]
Update: We created a workgroup on the Discord Community server for this. Please join here: https://discord.gg/w4GUrKvBTT

**Comment by fsbraun** [comment]
### AI Workgroup Update: Proposal to Explore an MCP Server for Django CMS

Following recent AI workgroup discussions, @mj8arrett and I would like to propose an investigation into creating an MCP (Model Context Protocol) server for Django CMS.

#### Goal (initial phase)
Enable an LLM to act as an assistive agent (“copilot” mode) for django CMS content, operating through a chat-based interface. The intent is supervised, intent-driven assistance with common editorial tasks (e.g. creating, rewriting, organizing content). It will be a tool for editors.

#### Key characteristics of the proposed approach

* The LLM would interact with Django CMS via an MCP server, rather than integrating directly into the admin UI.
CMS content would be exposed to the LLM through semantic, editorial projections, not raw database models or templates.
This aligns with a “copilot” model: assistive, auditable, and human-in-the-loop.
Initial scope

* Read and reason over CMS content, mostly text-based.
* Propose and apply safe, well-defined editorial changes to drafts.
* No automatic publishing; publishing remains an explicit human action.
#### Execution plan (step 1)

* Start with a separate chat application that supports MCP (e.g. Claude Desktop) rather than embedding this directly into the Django CMS UI.
* This allows rapid experimentation and iteration without impacting core CMS UX.
* The MCP server would be Django CMS–specific, but modular and optional.
#### Next steps (proposed)

* Validate feasibility and architecture of an MCP server for Django CMS.
* Identify which domain concepts and actions should be exposed in a first iteration.
* Assess risks, permissions, and audit requirements.
* Report back with findings and a recommendation on whether to proceed with a prototype.
* This is an exploratory proposal, intended to better understand how modern LLM tooling could integrate with Django CMS in a way that is safe, maintainable, and aligned with editorial workflows.

**Reply by fsbraun** [reply]
First insight: We need to run Django using asgi (i.e., asynchronous) as opposed to using wsgi if we want to benefit from the MCP Python SDK. django CMS 5.1 can [fully support](https://github.com/django-cms/django-cms/pull/8460) this (actually, django CMS 4.1 does as long as you do not use app hooks).

**Comment by mj8arrett** [comment]
For my part, I have been experimenting with a WordPress site, as there are several plugins that allow access via MCP.

I have spent about three hours playing around with it, and the results are very impressive.  Without using the WordPress editor at all, I have managed to do the following within Claude Code:

- Add/edit/delete pages/posts
- Copy content from an external URL and create a new page
- Import imagery from an external URL into the Media section
- Checked images for missing alt tags
- Add/edit/delete categories for article types
- Create 10 fictional news articles with dates over the last 12 months and upload/publish them on the site
- Created a tagging architecture from the article content and applied tags to all articles
- I even managed to get it to modify the theme to add the category label to the listing page

You can view the site here:

https://mcp.instawp.site/

Keep in mind, we might break it periodically as we play around with the more advanced functionality.

If we can create this kind of functionality in djangoCMS, then I think it is a game-changer for:

- Editors using AI to create and manage content
- Agency looking to automate routine tasks like content migration, SEO, Accessibility etc.

This also opens the opportunity to combine access/data from other services with MCP servers, such as Google Analytics, Search Console, Looker Studio, etc.

Below is a list of the Wordpress functions that are supported by the MCP server.  Each can be configured in the plugin to Always allow/Request permission/Blocked:
```
assign_terms_to_content
create_content
create_term
delete_content
delete_term
discover_content_types
discover_taxonomies
find_content_by_url
get_content
get_content_by_slug
get_content_terms
get_taxonomy
get_term
list_content
list_terms
media_attach
media_metadata
media_upload
patch_content
plugin_files
plugin_info
plugin_operations
theme_files
theme_info
theme_operations
update_content
update_term
```
I will continue my research next week and report back on any further findings/ideas.
